packages/zipkin/index.d.ts

```diff
@@ -80,7 +80,7 @@ declare namespace zipkin {
 
       map<V>(fn: (value: T) => V): IOption<V>;
       ifPresent<V>(fn: (value: T) => V): IOption<V>;
-      flatMap<V>(fn: (value: T) => V): IOption<V>;
+      flatMap<V>(fn: (value: T) => IOption<V>): IOption<V>;
```

Comment:
Thanks. I'm totally fine with making the types match the reality of the implementation.

One observation here is that returning a union makes this a bit more cumbersome for callers.  If all they know is that they have an IOption, they'll have to check flatMap's return value.  For example, TS users will be prevented from chaining any IOption methods after flatMap: `anOption.flatMap(f).map(g)` will be disallowed.

Reply:
I totally agree with you about the cumbersome and mess public API. As you said below, there are two ways of solving it, either changing TS definition or changing actual implementation. But changing the actual implementation is hard because we don't know the effects to functions using it. I also favor changing the actual implementation if it is not consistent but I would ask @jcchavezs for a recommendation.